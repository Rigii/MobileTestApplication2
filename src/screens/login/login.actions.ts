import { USER_EMAIL } from '../../constants/action.constants';
import { string } from 'prop-types';

export const setEmail = (email: string) => {
    return {
        type: USER_EMAIL,
        payload: email
    }
}