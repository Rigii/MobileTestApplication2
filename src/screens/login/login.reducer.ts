export const USER_EMAIL = "USER_EMAIL"

interface InitialStateTypes{
    email: string
}

export const initialState: InitialStateTypes = {
    email: ''
}

export const loginReducer = (state = initialState, action: any) => {
    switch (action.type) {
        case USER_EMAIL: return { ...state, email: action.payload };
        default: return state;
    }
}

