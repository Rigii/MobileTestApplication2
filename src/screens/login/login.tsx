import React, {useState} from 'react';
import {connect} from 'react-redux';
import {
  TextInput,
  View,
  Text,
  TouchableOpacity,
  StyleSheet,
  AsyncStorage,
} from 'react-native';
import {NavigationInjectedProps} from 'react-navigation';
import {ROUTES} from '../../constants/routes';
import {COLORS} from '../../constants/theme';
import {STRINGS} from '../../constants/strings';
import {setEmail} from './login.actions';
import {getTodoList} from '../todo-list/todo.actions';
import {TypesTodoList} from '../todo-list/todo.reducer';
import { fireEvent, render, wait } from '@testing-library/react-native';

interface LoginProps  extends NavigationInjectedProps { 
  getTodoList: (parsedData: string) => void;
  
}


// TODO: add snapshot test here, shallow test
// mock props method with jest.fn() and test that it was called
export const LoginScreenComp = (props: LoginProps) => {
  const [email, setEmail] = useState('');

  const setEmailGoToList = () => {
    if (email.length === 0) {
      return;
    }
    props.setEmail(email);
    getAsyncStorageData(email);
    props.navigation.navigate(ROUTES.TodoList);
  };

  const getAsyncStorageData = async (email: string) => {
    try {
      const data = await AsyncStorage.getItem(email);
      if (!data) {
        return;
      }
      let parsedData = JSON.parse(data);
      props.getTodoList(parsedData);
    } catch (error) {
      console.log(
        'Getting async storage data error, or such storage does not exist.',
      );
    }
  };

  return (
    <View style={styles.login_view}>
      <Text>{STRINGS.LOGIN.head_text}</Text>
      <TextInput
        style={[styles.input, styles.login_comp]}
        onChangeText={setEmail}
        testID="loginInput"
      />
      <TouchableOpacity
        style={[styles.but, styles.login_comp]}
        onPress={setEmailGoToList}
        testID="loginButton"
        >
        <Text style={styles.text}>{STRINGS.LOGIN.actionButton}</Text>
      </TouchableOpacity>
    </View>
  );
};

const mapDispatchToProps = {
  setEmail,
  getTodoList,
};

export const LoginScreen = connect(null, mapDispatchToProps)(LoginScreenComp);


// test('examples of some things', async () => {
//   const { getByTestId, getByText, queryByTestId, baseElement } = render(<LoginScreenComp />);
//   const famousWomanInHistory = 'Ada Lovelace';
 
//   const input = getByTestId('input');
//   fireEvent.changeText(input, famousWomanInHistory);
 
//   const button = getByText('Print Username');
//   fireEvent.press(button);
 
//   await wait(() => expect(queryByTestId('printed-username')).toBeTruthy());
 
//   expect(getByTestId('printed-username').props.children).toBe(famousWomanInHistory);
//   expect(baseElement).toMatchSnapshot();
// });



const styles = StyleSheet.create({
  login_view: {
    flex: 1,
    alignItems: 'center',
    justifyContent: 'center',
    alignContent: 'center',
    margin: 15,
  },
  login_comp: {
    marginTop: 20,
  },
  input: {
    width: 350,
    height: 40,
    borderColor: COLORS.input_border,
    borderWidth: 1,
    borderRadius: 5,
  },
  text: {
    alignSelf: 'center',
    fontSize: 14,
  },
  but: {
    flex: 0,
    backgroundColor: COLORS.main,
    borderRadius: 5,
    padding: 15,
    paddingHorizontal: 20,
    alignSelf: 'center',
  },
});
