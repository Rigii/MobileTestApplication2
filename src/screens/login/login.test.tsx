import * as React from 'react';
import {LoginScreenComp} from './login';
import {render, fireEvent} from '@testing-library/react-native';


describe('Login Screen Component', () => {
  it('Should match to snapshot', () => {
    const props = {
      getTodoList: jest.fn(),
      setEmail: jest.fn(),
    };
    const {asJSON} = render(<LoginScreenComp {...props} />);

    expect(asJSON()).toMatchSnapshot(); // snapshot settings 
  });

  it('Should contain email input and submit button', () => {
    const props = {
        getTodoList: jest.fn(),
        setEmail: jest.fn(),
    };
    const {getByTestId} = render(<LoginScreenComp {...props} />); // return pseudorender marcup structure 

    const input = getByTestId('loginInput'); // getting feld
    const button = getByTestId('loginButton');

    expect(input).toBeTruthy(); // checking is input is exist
    expect(button).toBeTruthy()
  });

  it('Should call setEmail function with user email after button is pressed', () => {
    const props = {
        getTodoList: jest.fn(),
        setEmail: jest.fn(),
        navigation: {navigate: jest.fn()}
    };
    const {getByTestId} = render(<LoginScreenComp {...props} />);

    const input = getByTestId('loginInput');
    const button = getByTestId('loginButton');

      fireEvent.changeText(input, 'famousWomanInHistory')
      fireEvent.press(button)

      expect(props.setEmail).toHaveBeenCalledWith('famousWomanInHistory') // is function setEmail was called with mocked arguments 
  })
});
