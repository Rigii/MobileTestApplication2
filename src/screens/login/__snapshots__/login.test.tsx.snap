// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Login Screen Component Should match to snapshot 1`] = `
<View
  collapsable={true}
  pointerEvents="box-none"
  style={
    Object {
      "flex": 1,
    }
  }
>
  <View
    style={
      Object {
        "alignContent": "center",
        "alignItems": "center",
        "flex": 1,
        "justifyContent": "center",
        "margin": 15,
      }
    }
  >
    <Text>
      LOGIN SCREN
    </Text>
    <TextInput
      allowFontScaling={true}
      rejectResponderTermination={true}
      style={
        Array [
          Object {
            "borderColor": "gray",
            "borderRadius": 5,
            "borderWidth": 1,
            "height": 40,
            "width": 350,
          },
          Object {
            "marginTop": 20,
          },
        ]
      }
      testID="loginInput"
      underlineColorAndroid="transparent"
    />
    <View
      accessible={true}
      focusable={true}
      isTVSelectable={true}
      style={
        Object {
          "alignSelf": "center",
          "backgroundColor": "orange",
          "borderRadius": 5,
          "flex": 0,
          "marginTop": 20,
          "opacity": 1,
          "padding": 15,
          "paddingHorizontal": 20,
        }
      }
      testID="loginButton"
    >
      <Text
        style={
          Object {
            "alignSelf": "center",
            "fontSize": 14,
          }
        }
      >
        GO TO PROFILE
      </Text>
    </View>
  </View>
</View>
`;
